packages/xen-api/src/index.js

```diff
@@ -726,29 +735,40 @@ export class Xapi extends EventEmitter {
   // 3. the session is renewed
   // 4. the second call fails with SESSION_INVALID which leads to a new
   //    unnecessary renewal
+  _sessionOpenRetryOptions = {
+    tries: 2,
+    when: { code: 'HOST_IS_SLAVE' },
+    onRetry: error => {
+      this._setUrl({ ...this._url, hostname: error.params[0] })
+    },
+  }
   _sessionOpen = coalesceCalls(this._sessionOpen)
   async _sessionOpen() {
-    const { user, password } = this._auth
-    const params = [user, password]
-    this._sessionId = await pRetry(
-      () => this._interruptOnDisconnect(this._call('session.login_with_password', params)),
-      {
-        tries: 2,
-        when: { code: 'HOST_IS_SLAVE' },
-        onRetry: error => {
-          this._setUrl({ ...this._url, hostname: error.params[0] })
-        },
-      }
-    )
+    const { user, password, sessionId } = this._auth
+
+    this._sessionId = sessionId
+
+    if (sessionId === undefined) {
+      const params = [user, password]
+      this._sessionId = await pRetry(
+        () => this._interruptOnDisconnect(this._call('session.login_with_password', params)),
+        this._sessionOpenRetryOptions
+      )
+    }
 
     const oldPoolRef = this._pool?.$ref
 
     // Similar to `(await this.getAllRecords('pool'))[0]` but prevents a
     // deadlock in case of error due to a pRetry calling _sessionOpen again
-    const pools = await this._call('pool.get_all_records', [this._sessionId])
+    const pools = await pRetry(
+      () => this._call('pool.get_all_records', [this._sessionId]),
+      this._sessionOpenRetryOptions
+    )
     const poolRef = Object.keys(pools)[0]
     this._pool = this._wrapRecord('pool', poolRef, pools[poolRef])
 
+    this.emit('sessionId', this._sessionId)
```

Comment:
Why?

Reply:
So that when we use credentials to connect, we can still grab the session ID when the session is renewed for some reason. This will be useful in XO Lite when the user logs in for the first time (with the password): if the session ID changes during that first session, we can still save the latest (ie valid) session ID in the cookies.

Reply:
This could have been a separate commit or PR, but let's not bother :slightly_smiling_face: